File: Code-By-Language/Kotlin/March-1-2018.kt
//****************
//*  1. Two Sum  *
//****************

//I know it's meant to be similar to Java, it is based off of it, but I struggle to see the actual use of it as an alternative

//O(n) solution
class Solution 
{
    fun twoSum(nums: IntArray, target: Int): IntArray 
    {
        var temp:Int = 0
        val returnArr: IntArray = intArrayOf(0,0)
        val myHashMap = HashMap<Int, Int>()
        
        for (i in nums.indices)
        {
            temp = target - nums[i]
            
            if(myHashMap.contains(temp))
            {
                returnArr[0] = myHashMap.get(temp)!!
                returnArr[1] = i
                return returnArr
            }

            myHashMap.put(nums[i],i)
        }
        
        return returnArr
    }
}

File: 2018/March/March-1/March-1-2018.scala
//****************
//*  1. Two Sum  *
//****************

//Better than Swift (screw those Optionals), but potentially very annoying

//O(n) solution
object Solution 
{
    def twoSum(nums: Array[Int], target: Int): Array[Int] = 
    {
        var returnArr = new Array[Int](2)
        returnArr(0) = 0
        returnArr(1) = 0
        var temp = 0
        var numsLen = nums.length
        var i = 0
        var myHashMap = scala.collection.mutable.Map(nums(0) -> 0)
        
        for (i<-1 to (numsLen-1))
        {
            temp = target - nums(i)
            
            if(myHashMap.contains(temp))
            {
                returnArr(0) = myHashMap(temp)
                returnArr(1) = i
            }
            
            myHashMap += (nums(i)->i)
        }
        
        returnArr
    }
}
